wings/market/coinbase_pro_market.pyx

```diff
@@ -0,0 +1,816 @@
+        cdef:
+            int64_t last_tick = <int64_t>(self._last_timestamp / self._poll_interval)
+            int64_t current_tick = <int64_t>(timestamp / self._poll_interval)
+
+        MarketBase.c_tick(self, timestamp)
+        if current_tick > last_tick:
+            if not self._poll_notifier.is_set():
+                self._poll_notifier.set()
+        self._last_timestamp = timestamp
+
+    async def _http_client(self) -> aiohttp.ClientSession:
+        if self._shared_client is None:
+            self._shared_client = aiohttp.ClientSession()
+        return self._shared_client
+
+    async def _api_request(self,
+                           http_method: str,
+                           path_url: str = None,
+                           url: str = None,
+                           data: Optional[Dict[str, any]] = None) -> Dict[str, any]:
+        assert path_url is not None or url is not None
+
+        url = f"{self.COINBASE_API_ENDPOINT}{path_url}" if url is None else url
+        data_str = "" if data is None else json.dumps(data)
+        headers = self.coinbase_auth.get_headers(http_method, path_url, data_str)
+
+        client = await self._http_client()
+        async with client.request(http_method,
+                                  url=url, timeout=self.API_CALL_TIMEOUT, data=data_str, headers=headers) as response:
+            data = await response.json()
+            if response.status != 200:
+                raise IOError(f"Error fetching data from {url}. HTTP status is {response.status}. {data}")
+            return data
+
+    async def _update_balances(self):
+        cdef:
+            dict account_info
+            list balances
+            str asset_name
+            set local_asset_names = set(self._account_balances.keys())
+            set remote_asset_names = set()
+            set asset_names_to_remove
+
+        path_url = "/accounts"
+        account_balances = await self._api_request('get', path_url=path_url)
+
+        for balance_entry in account_balances:
+            asset_name = balance_entry["currency"]
+            balance = Decimal(balance_entry["balance"])
+            self._account_balances[asset_name] = balance
+            remote_asset_names.add(asset_name)
+
+        asset_names_to_remove = local_asset_names.difference(remote_asset_names)
+        for asset_name in asset_names_to_remove:
+            del self._account_balances[asset_name]
+
+    async def _update_eth_tx_status(self):
+        in_flight_deposits = [d for d in self._in_flight_deposits.values() if not d.has_tx_receipt]
+        tx_hash_to_deposit_map = dict((d.tx_hash, d) for d in in_flight_deposits)
+
+        for d in in_flight_deposits:
+            receipt = self._w3.eth.getTransactionReceipt(d.tx_hash)
+            if receipt is None or receipt.blockHash is None:
+                continue
+            if receipt.status == 0:
+                d.has_tx_receipt = True
+                self.c_did_fail_tx(d.tracking_id)
```

Comment:
there is no event for failed tx? similar to below

Reply:
@chowryan tx failure event is thrown in c_did_fail_tx.